File: frontend/src/services/sentryConfig.ts
// frontend/src/services/sentryConfig.ts

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';
import { CaptureConsole } from '@sentry/integrations';

interface SentryConfigOptions {
  dsn?: string;
  environment?: string;
  release?: string;
  tracesSampleRate?: number;
}

class SentryService {
  private initialized: boolean = false;
  private dsn: string | undefined;
  private environment: string;

  constructor() {
    this.dsn = import.meta.env.VITE_SENTRY_DSN;
    this.environment = import.meta.env.VITE_ENVIRONMENT || 'development';
  }

  /**
   * Initialize Sentry for error and performance monitoring
   */
  initialize(options?: SentryConfigOptions): boolean {
    // Use provided DSN or fall back to environment variable
    const dsn = options?.dsn || this.dsn;

    if (!dsn) {
      console.warn('Sentry DSN not configured - monitoring disabled');
      return false;
    }

    if (this.initialized) {
      console.info('Sentry already initialized');
      return true;
    }

    try {
      Sentry.init({
        dsn,
        environment: options?.environment || this.environment,
        release: options?.release || import.meta.env.VITE_APP_VERSION || '1.0.0',

        // Performance Monitoring
        integrations: [
          new BrowserTracing({
            // Set sampling rate for performance monitoring
            tracingOrigins: [
              'localhost',
              /^https:\/\/marta-eta\.vercel\.app/,
              /^https:\/\/.*\.supabase\.co/,
            ],
            // Capture interactions
            routingInstrumentation: Sentry.reactRouterV6Instrumentation(
              React.useEffect,
              useLocation,
              useNavigationType,
              createRoutesFromChildren,
              matchRoutes
            ),
          }),
          // Capture console errors
          new CaptureConsole({
            levels: ['error', 'warn'],
          }),
        ],

        // Performance sampling
        tracesSampleRate: options?.tracesSampleRate || 0.1,

        // Session replay
        replaysSessionSampleRate: 0.1,
        replaysOnErrorSampleRate: 1.0,

        // Error filtering
        beforeSend: (event, hint) => {
          // Filter out certain errors
          if (event.exception) {
            const error = hint.originalException;

            // Don't send network errors in development
            if (this.environment === 'development' && error?.name === 'NetworkError') {
              return null;
            }

            // Filter out expected errors
            if (error?.message?.includes('ResizeObserver loop limit exceeded')) {
              return null;
            }
          }

          // Remove sensitive data
          if (event.request) {
            // Remove authorization headers
            if (event.request.headers) {
              delete event.request.headers['Authorization'];
              delete event.request.headers['Cookie'];
            }
          }

          return event;
        },

        // Don't send personally identifiable information
        sendClientReports: false,

        // Breadcrumb configuration
        maxBreadcrumbs: 50,

        // Ignore certain errors
        ignoreErrors: [
          // Browser extensions
          'top.GLOBALS',
          // Random network errors
          'Network request failed',
          'NetworkError',
          'Failed to fetch',
          // Resize observer errors
          'ResizeObserver loop limit exceeded',
          // Non-error promise rejections
          'Non-Error promise rejection captured',
        ],

        // Ignore transactions from certain URLs
        ignoreTransactions: [
          '/health',
          '/metrics',
        ],
      });

      this.initialized = true;
      console.info(`Sentry initialized for ${this.environment} environment`);
      return true;
    } catch (error) {
      console.error('Failed to initialize Sentry:', error);
      return false;
    }
  }

  /**
   * Capture an exception with optional context
   */
  captureException(error: Error, context?: Record<string, any>): void {
    if (!this.initialized) return;

    if (context) {
      Sentry.withScope((scope) => {
        Object.keys(context).forEach((key) => {
          scope.setContext(key, context[key]);
        });
        Sentry.captureException(error);
      });
    } else {
      Sentry.captureException(error);
    }
  }

  /**
   * Capture a message with optional context
   */
  captureMessage(
    message: string,
    level: Sentry.SeverityLevel = 'info',
    context?: Record<string, any>
  ): void {
    if (!this.initialized) return;

    if (context) {
      Sentry.withScope((scope) => {
        Object.keys(context).forEach((key) => {
          scope.setContext(key, context[key]);
        });
        Sentry.captureMessage(message, level);
      });
    } else {
      Sentry.captureMessage(message, level);
    }
  }

  /**
   * Add breadcrumb for tracking user actions
   */
  addBreadcrumb(breadcrumb: Sentry.Breadcrumb): void {
    if (!this.initialized) return;
    Sentry.addBreadcrumb(breadcrumb);
  }

  /**
   * Set user context
   */
  setUser(user: Sentry.User | null): void {
    if (!this.initialized) return;
    Sentry.setUser(user);
  }

  /**
   * Set additional context
   */
  setContext(key: string, context: any): void {
    if (!this.initialized) return;
    Sentry.setContext(key, context);
  }

  /**
   * Set tags for categorization
   */
  setTag(key: string, value: string): void {
    if (!this.initialized) return;
    Sentry.setTag(key, value);
  }

  /**
   * Start a transaction for performance monitoring
   */
  startTransaction(name: string, op: string = 'navigation'): any {
    if (!this.initialized) return null;
    return Sentry.startTransaction({ name, op });
  }

  /**
   * Profile a React component
   */
  profileComponent<P extends object>(
    Component: React.ComponentType<P>,
    name?: string
  ): React.ComponentType<P> {
    if (!this.initialized) return Component;
    return Sentry.withProfiler(Component, { name });
  }

  /**
   * Create error boundary component
   */
  createErrorBoundary(fallback: React.ComponentType<any>, showDialog: boolean = false) {
    return Sentry.ErrorBoundary({ fallback, showDialog });
  }
}

// Create singleton instance
export const sentryService = new SentryService();

// React hooks for Sentry
export const useSentryUser = (user: any) => {
  React.useEffect(() => {
    if (user) {
      sentryService.setUser({
        id: user.id,
        email: user.email,
        username: user.username,
      });
    } else {
      sentryService.setUser(null);
    }
  }, [user]);
};

export const useSentryError = () => {
  return {
    captureException: (error: Error, context?: Record<string, any>) =>
      sentryService.captureException(error, context),
    captureMessage: (message: string, level?: Sentry.SeverityLevel, context?: Record<string, any>) =>
      sentryService.captureMessage(message, level, context),
  };
};

// Error boundary wrapper component
export const SentryErrorBoundary: React.FC<{
  children: React.ReactNode;
  fallback?: React.ComponentType<any>;
}> = ({ children, fallback }) => {
  const ErrorBoundary = sentryService.createErrorBoundary(
    fallback || DefaultErrorFallback,
    false
  );

  return <ErrorBoundary>{children}</ErrorBoundary>;
};

// Default error fallback component
const DefaultErrorFallback: React.FC<{ error?: Error }> = ({ error }) => (
  <div className="min-h-screen flex items-center justify-center bg-gray-100">
    <div className="bg-white p-8 rounded-lg shadow-md max-w-md w-full">
      <h2 className="text-2xl font-bold text-red-600 mb-4">Something went wrong</h2>
      <p className="text-gray-600 mb-4">
        We're sorry, but something unexpected happened. The error has been reported to our team.
      </p>
      {error && (
        <details className="mb-4">
          <summary className="cursor-pointer text-sm text-gray-500">Error details</summary>
          <pre className="mt-2 text-xs bg-gray-100 p-2 rounded overflow-auto">
            {error.message}
          </pre>
        </details>
      )}
      <button
        onClick={() => window.location.reload()}
        className="w-full bg-blue-600 text-white py-2 px-4 rounded hover:bg-blue-700"
      >
        Reload Page
      </button>
    </div>
  </div>
);

// Import required React Router dependencies
import React from 'react';
import {
  useLocation,
  useNavigationType,
  createRoutesFromChildren,
  matchRoutes,
} from 'react-router-dom';

export default sentryService;

File: src/services/sentry_config.py
# src/services/sentry_config.py

import os
import logging
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.logging import LoggingIntegration
from sentry_sdk.integrations.redis import RedisIntegration
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

class SentryConfig:
    """
    Sentry monitoring configuration for MARTA Transit Analytics Platform.
    """

    def __init__(self):
        self.dsn = os.getenv("SENTRY_DSN")
        self.environment = os.getenv("ENVIRONMENT", "development")
        self.release = os.getenv("APP_VERSION", "1.0.0")
        self.traces_sample_rate = float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "0.1"))
        self.profiles_sample_rate = float(os.getenv("SENTRY_PROFILES_SAMPLE_RATE", "0.1"))
        self.initialized = False

    def initialize(self, app_name: str = "marta-backend") -> bool:
        """
        Initialize Sentry monitoring.

        Args:
            app_name: Name of the application

        Returns:
            True if successfully initialized
        """
        if not self.dsn:
            logger.warning("Sentry DSN not configured - monitoring disabled")
            return False

        if self.initialized:
            logger.info("Sentry already initialized")
            return True

        try:
            # Configure logging integration
            logging_integration = LoggingIntegration(
                level=logging.INFO,
                event_level=logging.ERROR
            )

            # Initialize Sentry
            sentry_sdk.init(
                dsn=self.dsn,
                environment=self.environment,
                release=self.release,
                traces_sample_rate=self.traces_sample_rate,
                profiles_sample_rate=self.profiles_sample_rate,
                integrations=[
                    FastApiIntegration(transaction_style="endpoint"),
                    SqlalchemyIntegration(),
                    logging_integration,
                    RedisIntegration(),
                ],
                # Performance monitoring
                enable_tracing=True,

                # Session tracking
                release_health=True,

                # Error filtering
                before_send=self._before_send,
                before_send_transaction=self._before_send_transaction,

                # Additional options
                attach_stacktrace=True,
                send_default_pii=False,  # Don't send personally identifiable information
                max_breadcrumbs=50,
                debug=self.environment == "development",

                # Custom tags
                tags={
                    "app": app_name,
                    "service": "marta-transit",
                    "component": "backend"
                }
            )

            # Set user context if available
            self._set_user_context()

            # Set additional context
            sentry_sdk.set_context("app_info", {
                "name": app_name,
                "version": self.release,
                "environment": self.environment
            })

            self.initialized = True
            logger.info(f"Sentry initialized for {app_name} in {self.environment} environment")
            return True

        except Exception as e:
            logger.error(f"Failed to initialize Sentry: {e}")
            return False

    def capture_exception(self, exception: Exception, context: Optional[Dict[str, Any]] = None):
        """
        Capture exception with optional context.

        Args:
            exception: The exception to capture
            context: Additional context to attach
        """
        if not self.initialized:
            return

        try:
            if context:
                with sentry_sdk.push_scope() as scope:
                    for key, value in context.items():
                        scope.set_context(key, value)
                    sentry_sdk.capture_exception(exception)
            else:
                sentry_sdk.capture_exception(exception)
        except Exception as e:
            logger.error(f"Failed to capture exception in Sentry: {e}")

    def capture_message(self, message: str, level: str = "info", context: Optional[Dict[str, Any]] = None):
        """
        Capture a message with optional context.

        Args:
            message: The message to capture
            level: Log level (debug, info, warning, error, fatal)
            context: Additional context to attach
        """
        if not self.initialized:
            return

        try:
            if context:
                with sentry_sdk.push_scope() as scope:
                    for key, value in context.items():
                        scope.set_context(key, value)
                    sentry_sdk.capture_message(message, level=level)
            else:
                sentry_sdk.capture_message(message, level=level)
        except Exception as e:
            logger.error(f"Failed to capture message in Sentry: {e}")

    def add_breadcrumb(self, message: str, category: str = "custom", level: str = "info", data: Optional[Dict] = None):
        """
        Add breadcrumb for tracking user actions.

        Args:
            message: Breadcrumb message
            category: Category of the breadcrumb
            level: Log level
            data: Additional data
        """
        if not self.initialized:
            return

        try:
            sentry_sdk.add_breadcrumb(
                message=message,
                category=category,
                level=level,
                data=data or {}
            )
        except Exception as e:
            logger.error(f"Failed to add breadcrumb: {e}")

    def set_user(self, user_id: str, email: Optional[str] = None, username: Optional[str] = None):
        """
        Set user context for error tracking.

        Args:
            user_id: Unique user identifier
            email: User email
            username: Username
        """
        if not self.initialized:
            return

        try:
            sentry_sdk.set_user({
                "id": user_id,
                "email": email,
                "username": username
            })
        except Exception as e:
            logger.error(f"Failed to set user context: {e}")

    def start_transaction(self, name: str, op: str = "function") -> Optional[Any]:
        """
        Start a performance transaction.

        Args:
            name: Transaction name
            op: Operation type

        Returns:
            Transaction object or None
        """
        if not self.initialized:
            return None

        try:
            return sentry_sdk.start_transaction(name=name, op=op)
        except Exception as e:
            logger.error(f"Failed to start transaction: {e}")
            return None

    def _before_send(self, event: Dict, hint: Dict) -> Optional[Dict]:
        """
        Filter or modify events before sending to Sentry.

        Args:
            event: The event to be sent
            hint: Additional information about the event

        Returns:
            Modified event or None to drop it
        """
        # Filter out certain errors
        if "exc_info" in hint:
            exc_type, exc_value, tb = hint["exc_info"]

            # Don't send certain expected errors
            if exc_type.__name__ in ["KeyboardInterrupt", "SystemExit"]:
                return None

            # Filter out 404 errors in production
            if self.environment == "production" and hasattr(exc_value, "status_code"):
                if exc_value.status_code == 404:
                    return None

        # Remove sensitive data
        if "request" in event:
            request = event["request"]
            # Remove authorization headers
            if "headers" in request:
                request["headers"] = {
                    k: v for k, v in request["headers"].items()
                    if k.lower() not in ["authorization", "cookie", "x-api-key"]
                }
            # Remove sensitive query parameters
            if "query_string" in request:
                # Parse and filter query string if needed
                pass

        return event

    def _before_send_transaction(self, event: Dict, hint: Dict) -> Optional[Dict]:
        """
        Filter or modify transaction events before sending to Sentry.

        Args:
            event: The transaction event
            hint: Additional information

        Returns:
            Modified event or None to drop it
        """
        # Filter out health check endpoints
        if "transaction" in event:
            transaction_name = event["transaction"]
            if transaction_name in ["/health", "/", "/metrics"]:
                return None

        return event

    def _set_user_context(self):
        """Set user context from environment or request."""
        # This would typically be set per request in middleware
        # Example implementation for default context
        user_id = os.getenv("DEFAULT_USER_ID")
        if user_id:
            self.set_user(user_id)

# Create singleton instance
sentry_config = SentryConfig()

# Middleware for FastAPI
class SentryMiddleware:
    """
    FastAPI middleware for Sentry integration.
    """

    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        # Extract request information
        path = scope.get("path", "")
        method = scope.get("method", "")

        # Start transaction
        transaction = sentry_config.start_transaction(
            name=f"{method} {path}",
            op="http.server"
        )

        if transaction:
            with transaction:
                # Add request information
                transaction.set_tag("http.method", method)
                transaction.set_tag("http.url", path)

                # Process request
                try:
                    await self.app(scope, receive, send)
                except Exception as e:
                    transaction.set_status("internal_error")
                    sentry_config.capture_exception(e, context={
                        "request": {
                            "method": method,
                            "path": path
                        }
                    })
                    raise
                else:
                    transaction.set_status("ok")
        else:
            await self.app(scope, receive, send)

# Decorators for function monitoring
def monitor_performance(op: str = "function"):
    """
    Decorator to monitor function performance.

    Args:
        op: Operation type for Sentry
    """
    def decorator(func):
        def wrapper(*args, **kwargs):
            transaction = sentry_config.start_transaction(
                name=f"{func.__module__}.{func.__name__}",
                op=op
            )

            if transaction:
                with transaction:
                    try:
                        result = func(*args, **kwargs)
                        transaction.set_status("ok")
                        return result
                    except Exception as e:
                        transaction.set_status("internal_error")
                        sentry_config.capture_exception(e)
                        raise
            else:
                return func(*args, **kwargs)

        return wrapper
    return decorator

def capture_errors(func):
    """
    Decorator to capture function errors.
    """
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except Exception as e:
            sentry_config.capture_exception(e, context={
                "function": f"{func.__module__}.{func.__name__}",
                "args": str(args)[:500],  # Limit size
                "kwargs": str(kwargs)[:500]
            })
            raise

    return wrapper